sycl/include/sycl/ext/oneapi/properties/properties.hpp

```diff
@@ -9,265 +9,180 @@
-      typename ExtractProperties<std::tuple<PropertiesTs...>>::
-          template ExtractedPropertiesT<PropertyValueTs...>;
-  template <typename... PropertyValueTs>
-  using ExtractedPropertiesT =
-      typename PrependTuple<PropertyT,
-                            NextExtractedPropertiesT<PropertyValueTs...>>::type;
-
-  template <typename... PropertyValueTs>
-  static constexpr ExtractedPropertiesT<PropertyValueTs...>
-  Extract(std::tuple<PropertyValueTs...> PropertyValues) {
-    PropertyT ThisExtractedProperty = std::get<PropertyT>(PropertyValues);
-    NextExtractedPropertiesT<PropertyValueTs...> NextExtractedProperties =
-        ExtractProperties<std::tuple<PropertiesTs...>>::template Extract<
-            PropertyValueTs...>(PropertyValues);
-    return std::tuple_cat(std::tuple<PropertyT>{ThisExtractedProperty},
-                          NextExtractedProperties);
-  }
-};
-
+template <class V> using make_entry_t = typename make_entry<V>::type;
+// create property map from properties type
+template <class P>
+using property_map = mp11::mp_transform<make_entry_t, mp11::mp_front<P>>;
+template <class T, class = void>
+struct is_empty_or_incomplete : std::true_type {};
+template <class T>
+struct is_empty_or_incomplete<
+    T, std::enable_if_t<(sizeof(T) > 0) && !std::is_empty_v<T>>>
+    : std::false_type {};
+template <class T>
+constexpr bool is_empty_or_incomplete_v = is_empty_or_incomplete<T>::value;
+template <class T>
+using IsRuntimeProperty = mp11::mp_not<is_empty_or_incomplete<T>>;
 } // namespace detail
 
-template <typename PropertiesT> class properties {
-  static_assert(detail::IsTuple<PropertiesT>::value,
-                "Properties must be in a tuple.");
-  static_assert(detail::AllPropertyValues<PropertiesT>::value,
+// base case is only for invalid cases to suppress compiler errors about base
+// class storage of properties
+template <class L, class = void> class properties {
+  static_assert(detail::mp11::mp_all_of<L, std::is_class>(),
                 "Unrecognized property in property list.");
-  static_assert(detail::IsSorted<PropertiesT>::value,
-                "Properties in property list are not sorted.");
-  static_assert(detail::SortedAllUnique<PropertiesT>::value,
+  static_assert(detail::mp11::mp_is_set<L>(),
                 "Duplicate properties in property list.");
-  static_assert(detail::NoConflictingProperties<PropertiesT>::value,
-                "Conflicting properties in property list.");
-
-public:
-  template <typename... PropertyValueTs>
-  constexpr properties(PropertyValueTs... props)
-      : Storage(detail::ExtractProperties<StorageT>::Extract(
-            std::tuple<PropertyValueTs...>{props...})) {}
+};
 
-  template <typename PropertyT>
-  static constexpr std::enable_if_t<detail::IsProperty<PropertyT>::value, bool>
-  has_property() {
-    return detail::ContainsProperty<PropertyT, PropertiesT>::value;
-  }
+template <template <class...> class TL, class... V>
```

Comment:
This goes for a lot of the new template argument names: The names are both problematic because single capital letter names historically have a solid chance of macro conflicts. These names are also not very descriptive.

Reply:
I left `T` because it is anyhow used everwhere. But I replaced others.